[HLSL][DirectX] Add the Qdx-rootsignature-strip driver option

inbelic opened the pull request:
This pr adds the `Qstrip-rootsignature` as a `DXC` driver option.

To do so, this pr introduces the `BinaryModifyJobClass` as an `Action` to modify a produced object file before its final output.

Further, it registers `llvm-objcopy` as the tool to modify a produced `DXContainer` on the `HLSL` toolchain.

This allows us to specify the `Qstrip-rootsignature` option to `clang-dxc` which will invoke `llvm-objcopy` with a `--remove-section=RTS0` argument to implement its functionality.

Resolves: https://github.com/llvm/llvm-project/issues/150275.

inbelic committed:
add BinaryModifyJobClass

inbelic committed:
[HLSL] Add the `Qstrip-rootsignature` DXC driver option

llvm-beanz commented on clang/include/clang/Driver/Action.h:
```diff
@@ -76,9 +76,10 @@ class Action {
     StaticLibJobClass,
     BinaryAnalyzeJobClass,
     BinaryTranslatorJobClass,
+    BinaryModifyJobClass,
```
Probably better to align this with the specific tool you're running:
```suggestion
    ObjcopyJobClass,
```

That is consistent with most of the other toolchain jobs (e..g. LipoJobClass, DsymutilJobClass, LinkerWrapperJobClass, etc).

llvm-beanz commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4601,6 +4601,16 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
+      // llvm-objcopy expects a DXIL container, which can either be
+      // validated (in which case they are TY_DX_CONTAINER), or unvalidated
+      // (TY_OBJECT).
+      if (LastAction->getType() == types::TY_DX_CONTAINER ||
```
Should we be running this before validation rather than after? What does DXC do?

inbelic replied:
Yep, DXC will construct the final output blob before running the validator. So it would validate with the parts removed.

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -247,6 +247,30 @@ void tools::hlsl::MetalConverter::ConstructJob(
+    const char *Frs = Args.MakeArgString("--remove-section=RTS0");
+    CmdArgs.push_back(Frs);
+  }
+
```
Should we assert that `CmdArgs > 2` to ensure that we didn't just invoke objcopy for no reason?

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.h:
```diff
@@ -42,6 +42,19 @@ class LLVM_LIBRARY_VISIBILITY MetalConverter : public Tool {
+class LLVM_LIBRARY_VISIBILITY LLVMObjcopy : public Tool {
+public:
+  LLVMObjcopy(const ToolChain &TC)
+      : Tool("hlsl::LLVMObjcopy", "llvm-objcopy", TC) {}
```
I don't think the `hlsl` is relevant here. It's really just objcopy.
```suggestion
      : Tool("LLVMObjcopy", "llvm-objcopy", TC) {}
```

inbelic committed:
review: use terminology directed for objcopy

inbelic committed:
review: add assert

inbelic committed:
review: apply objcopy before validation

- DXC will invoke the validator after the final blob has been generated
(in this case after the part is removed)

farzonl commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -295,6 +295,32 @@ void tools::hlsl::MetalConverter::ConstructJob(
+                                      const ArgList &Args,
+                                      const char *LinkingOutput) const {
+
+  std::string ObjcopyPath = getToolChain().GetProgramPath("llvm-objcopy");
```
don't work much in the driver, should we be using a std:string here, I've noticed this file uses StringRef in other places.

llvm-beanz replied:
`GetProgramPath` returns a constructed string as a `std::string`, so you need to take ownership of the allocation. Using a StringRef to a returned temporary would be bad.

llvm-beanz commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4647,6 +4647,16 @@ void Driver::BuildDefaultActions(Compilation &C, DerivedArgList &Args,
+      Action *LastAction = Actions.back();
+      // llvm-objcopy expects a DXIL container, which can either be
+      // validated (in which case they are TY_DX_CONTAINER), or unvalidated
+      // (TY_OBJECT).
```
Didn't we confirm this should only run on unvalidated containers?

inbelic committed:
review: improve comments

inbelic committed:
re-review: make this llvmobcopy hlsl specific

- following convention used from other toolchains (eg. `Linker`), each
toolchain will need to define a custom override of `ConstructJob` so
that it converts the applicable arguments accordingly

- since this will be toolchain specific, it is convention to prefix it
with the toolchain namespace (hlsl::)

inbelic commented on clang/lib/Driver/ToolChains/HLSL.h:
```diff
@@ -42,7 +42,21 @@ class LLVM_LIBRARY_VISIBILITY MetalConverter : public Tool {
+
 } // namespace hlsl
+
+class LLVM_LIBRARY_VISIBILITY LLVMObjcopy : public Tool {
+public:
+  LLVMObjcopy(const ToolChain &TC) : Tool("LLVMObjcopy", "llvm-objcopy", TC) {}
+
+  bool hasIntegratedCPP() const override { return false; }
+
+  void ConstructJob(Compilation &C, const JobAction &JA,
+                    const InputInfo &Output, const InputInfoList &Inputs,
+                    const llvm::opt::ArgList &TCArgs,
+                    const char *LinkingOutput) const override;
+};
```
While it is generic, the logic of `ConstructJob` is specific to just the `hlsl` toolchain.

It is convention for the other toolchains to each redefine their own version of the struct with their namespace prefixed.  See `Linker` etc.

So I have added back the `hlsl` namespace to this struct.

inbelic committed:
review: remove unused marshalling

inbelic committed:
review: clarify `isLastJob` check

inbelic committed:
self-review: little clean-up

inbelic committed:
review: correct job order and add test that validation occurs after
